### PR TITLE
add ContentFilteredTopic example.

### PR DESCRIPTION
**Pull request by fujitatomoya**
should be linked to ROS 2 ContentFitleredTopic tutorial. (https://github.com/ros2/design/pull/282)

related to https://github.com/ros2/demos/pull/557

Signed-off-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

### CODE REVIEW

**Review by fujitatomoya — COMMENTED** [review]
just trying to use all interfaces provided by subscription content filtering here. and dealing with exceptions that might be generated via methods since users are likely to `copy&paste` the code from here to somewhere else.

**Comment by fujitatomoya** [comment]
CC: @wjwwood @ivanpauno @iuhilnehc-ynos

File: rclcpp/topics/minimal_subscriber/README.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ This package contains a few different strategies for creating nodes which receiv
  * `static_wait_set_subscriber.cpp` uses a `rclcpp::StaticWaitSet` to wait and poll for data
  * `time_triggered_wait_set_subscriber.cpp` uses a `rclcpp::Waitset` and a timer to poll for data
    periodically
+ * `content_filtering.cpp` uses a ROS 2 ContentFilteredTopic
```
Comment (wjwwood) [inline rclcpp/topics/minimal_subscriber/README.md:11]:
```suggestion
 * `content_filtering.cpp` used the content filtering feature for Subscriptions
```

I don't believe we have a `ContentFlitleredTopic` class or anything like that.

File: rclcpp/topics/minimal_subscriber/README.md [hunk 2]
```diff
@@ -20,3 +21,5 @@ We provide multiple examples of different coding styles which achieve this behav
 The following examples `wait_set_subscriber.cpp`, `static_wait_set_subscriber.cpp` and `time_triggered_wait_set_subscriber.cpp` show how to use a subscription in a node using a `rclcpp` wait-set.
 This is not a common use case in ROS 2 so this is not the recommended strategy to  use by-default.
 This strategy makes sense in some specific situations, for example when the developer needs to have more control over callback order execution, to create custom triggering conditions or to use the timeouts provided by the  wait-sets.   
+
+The following example `content_filtering.cpp` show how to use ROS 2 ContentFilteredTopic.
```
Comment (wjwwood) [inline rclcpp/topics/minimal_subscriber/README.md:25]:
```suggestion
The example `content_filtering.cpp` shows how to use the content filtering feature for Subscriptions.
```

**Review by wjwwood — COMMENTED** [review]
The example could be more "minimal" for copy-paste, but I think it's fine as-is to show more of how it could be used.

lgtm, with some suggestions addressed.

**Comment by fujitatomoya** [comment]
@wjwwood thanks for the comment, i borrowed your suggestions.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16592/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11174/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16978/)

**Comment by fujitatomoya** [comment]
CI (address uncrustify errors):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16593/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11175/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16979/)

**Comment by fujitatomoya** [comment]
example output for each rmw implementation.

- publisher (same binary with `RMW_IMPLEMENTATION` setting)

```bash
ros2 run examples_rclcpp_minimal_publisher publisher_member_function
```

- rmw_fastrtps (default)

```bash
# ros2 run examples_rclcpp_minimal_subscriber subscriber_content_filtering
[INFO] [1650647654.909896684] [minimal_contentfiltering_subscriber]: Subscribed to topic "/topic" with content filtering
[INFO] [1650647654.909990922] [minimal_contentfiltering_subscriber]: Content filtering expression and parameter are "data = %0" and "'Hello, world! 10'"
[INFO] [1650647663.337609015] [minimal_contentfiltering_subscriber]: I heard: 'Hello, world! 10'
[INFO] [1650647663.338324998] [minimal_contentfiltering_subscriber]: Content filtering expression and parameter are "data = %0" and "'Hello, world! 20'"
[INFO] [1650647668.337454893] [minimal_contentfiltering_subscriber]: I heard: 'Hello, world! 20'
[INFO] [1650647668.338167047] [minimal_contentfiltering_subscriber]: Content filtering expression and parameter are "data = %0" and "'Hello, world! 30'"
...<snip>
```

- rmw_connextdds

```bash
# RMW_IMPLEMENTATION=rmw_connextdds ros2 run examples_rclcpp_minimal_subscriber subscriber_content_filtering
RTI Connext DDS Non-commercial license is for academic, research, evaluation and personal use only. USE FOR COMMERCIAL PURPOSES IS PROHIBITED. See RTI_LICENSE.TXT for terms. Download free tools at rti.com/ncl. License issued to Non-Commercial User license@rti.com For non-production use only.
Expires on 00-jan-00 See www.rti.com for more information.
[INFO] [1650647909.339308552] [minimal_contentfiltering_subscriber]: Subscribed to topic "/topic" with content filtering
[INFO] [1650647909.339387411] [minimal_contentfiltering_subscriber]: Content filtering expression and parameter are "data = %0" and "'Hello, world! 10'"
[INFO] [1650647926.630042121] [minimal_contentfiltering_subscriber]: I heard: 'Hello, world! 10'
[INFO] [1650647926.631364854] [minimal_contentfiltering_subscriber]: Content filtering expression and parameter are "data = %0" and "'Hello, world! 20'"
...<snip>
```

- rmw_cyclonedds

```bash
# RMW_IMPLEMENTATION=rmw_cyclonedds_cpp ros2 run examples_rclcpp_minimal_subscriber subscriber_content_filtering
[WARN] [1650647987.733083073] [minimal_contentfiltering_subscriber]: Content filter is not enabled since it's not supported
[INFO] [1650647990.514812843] [minimal_contentfiltering_subscriber]: I heard: 'Hello, world! 0'
[INFO] [1650647991.014601673] [minimal_contentfiltering_subscriber]: I heard: 'Hello, world! 1'
[INFO] [1650647991.514611925] [minimal_contentfiltering_subscriber]: I heard: 'Hello, world! 2'
[INFO] [1650647992.014454205] [minimal_contentfiltering_subscriber]: I heard: 'Hello, world! 3'
[INFO] [1650647992.514540176] [minimal_contentfiltering_subscriber]: I heard: 'Hello, world! 4'
[INFO] [1650647993.014640461] [minimal_contentfiltering_subscriber]: I heard: 'Hello, world! 5'
...<snip>
```

**Comment by fujitatomoya** [comment]
Note: cross-vendor content filtering works between rmw_fastrtps and rmw_connextdds using this example.

**Review by ivanpauno — APPROVED** [review]
LGTM!

File: rclcpp/topics/minimal_subscriber/README.md [hunk 3]
```diff
@@ -8,6 +8,7 @@ This package contains a few different strategies for creating nodes which receiv
  * `static_wait_set_subscriber.cpp` uses a `rclcpp::StaticWaitSet` to wait and poll for data
  * `time_triggered_wait_set_subscriber.cpp` uses a `rclcpp::Waitset` and a timer to poll for data
    periodically
+ * `content_filtering.cpp` used the content filtering feature for Subscriptions
```
Comment (wjwwood) [inline rclcpp/topics/minimal_subscriber/README.md:11]:
```suggestion
 * `content_filtering.cpp` uses the content filtering feature for Subscriptions
```

**Review by wjwwood — APPROVED** [review]

**Review by wjwwood — APPROVED** [review]

**Comment by fujitatomoya** [comment]
either @ivanpauno or @wjwwood , could you help to merge this? i do not have access permission for this repo.

**Comment by clalancette** [comment]
> either @ivanpauno or @wjwwood , could you help to merge this? i do not have access permission for this repo.

Let's hold off until next week; we are trying to branch for Humble right now.

**Comment by fujitatomoya** [comment]
sorry, forgot that. thanks!

**Comment by wjwwood** [comment]
Yeah, this is ready, just waiting on Humble to unblock.

**Comment by fujitatomoya** [comment]
@clalancette could you merge this into master? so that i can link this example from https://github.com/ros2/ros2_documentation/pull/2396 (release note).

**Comment by fujitatomoya** [comment]
@Mergifyio backport rolling humble

**Pull request merged**